### PR TITLE
Added ability to pass attributes scope using templates for Delete attributes and Save attributes nodes

### PR DESCRIPTION
**Pull request by irynamatveieva**
## Pull Request description

Added ability to pass attributes scope using templates for Delete attributes and Save attributes nodes.
PE PR: https://github.com/thingsboard/thingsboard-pe/pull/2475

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java [hunk 1]
```diff
@@ -180,6 +204,97 @@ void givenNotifyDeviceMdValue_whenSaveAndNotify_thenVerifyExpectedArgumentForNot
...
+
+        node.onMsg(ctx, msg);
+
+        verify(telemetryService).saveAndNotify(eq(ctx.getTenantId()), eq(msg.getOriginator()), eq(AttributeScope.SERVER_SCOPE),
+                anyList(), eq(true), eq(new TelemetryNodeCallback(ctx, msg)));
+        verify(ctx).tellSuccess(eq(msg));
+    }
+
+    @Test
+    void givenUseTemplateTrueAndStrangeDataKey_whenOnMsg_thenSaveAttributes() throws Exception {
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java:227]:
```suggestion
    void givenUseTemplateTrueAndSpongeCaseDataKey_whenOnMsg_thenSaveAttributes() throws Exception {
```
image

Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java:227]: you might call it "SpongeCase," drawing from the internet meme "Mocking SpongeBob," where text is typed with alternating uppercase and lowercase letters to convey a mocking tone.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java [hunk 2]
```diff
@@ -180,6 +204,97 @@ void givenNotifyDeviceMdValue_whenSaveAndNotify_thenVerifyExpectedArgumentForNot
...
+        config.setUpdateAttributesOnlyOnValueChange(false);
+        nodeConfiguration = new TbNodeConfiguration(JacksonUtil.valueToTree(config));
+        node.init(ctx, nodeConfiguration);
+
+        final String data = "{\"TestAttribute_2\": \"humidity\", \"TestAttribute_3\": \"voltage\", \"scopeInData\": \" SeRvEr_ScOpE \"}";
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_ATTRIBUTES_REQUEST, deviceId, TbMsgMetaData.EMPTY, data);
+
+        node.onMsg(ctx, msg);
+
+        verify(ctx).tellSuccess(eq(msg));
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java:239]: will be nice to have verify of:

verify(telemetryService).saveAndNotify(...) with eq(AttributeScope.SERVER_SCOPE)

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java [hunk 3]
```diff
@@ -180,6 +204,97 @@ void givenNotifyDeviceMdValue_whenSaveAndNotify_thenVerifyExpectedArgumentForNot
...
+    }
+
+    @Test
+    void givenUseTemplateTrueAndInvalidMetadataKey_whenOnMsg_thenThrowException() throws TbNodeException {
+        config.setUseAttributesScopeTemplate(true);
+        config.setScope("${scopeInMetadata}");
+        nodeConfiguration = new TbNodeConfiguration(JacksonUtil.valueToTree(config));
+        node.init(ctx, nodeConfiguration);
+
+        final Map<String, String> mdMap = Map.of("scopeMetadata", "SERVER_SCOPE");
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java:249]: maybe remove

final Map<String, String> mdMap = Map.of("scopeMetadata", "SERVER_SCOPE");

and use just TbMsgMetaData.EMPTY. The result should be the same.

File: common/data/src/main/java/org/thingsboard/server/common/data/AttributeScope.java [hunk 4]
```diff
@@ -29,4 +29,15 @@ public enum AttributeScope {
         this.id = id;
     }
 
+    public static AttributeScope parseFrom(String scope) {
```
Comment (ShvaykaD) [inline common/data/src/main/java/org/thingsboard/server/common/data/AttributeScope.java:32]: Please add test for this method. Test might be parameterized.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java [hunk 5]
```diff
@@ -69,12 +75,30 @@ class TbMsgAttributesNodeTest extends AbstractRuleNodeUpgradeTest {
...
     @BeforeEach
-    void setUp() {
+    void setUp() throws TbNodeException {
         tenantId = new TenantId(UUID.fromString("6c18691e-4470-4766-9739-aface71d761f"));
         deviceId = new DeviceId(UUID.fromString("b66159d7-c77e-45e8-bb41-a8f557f434c1"));
         node = spy(TbMsgAttributesNode.class);
+        ctx = mock(TbContext.class);
+        config = new TbMsgAttributesNodeConfiguration().defaultConfiguration();
+        nodeConfiguration = new TbNodeConfiguration(JacksonUtil.valueToTree(config));
+        node.init(ctx, nodeConfiguration);
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNodeTest.java:91]: Why you need to init in the setUp? I see that you do init almost in all your tests as well.

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgDeleteAttributesNodeTest.java [hunk 6]
```diff
@@ -127,6 +136,145 @@ void givenMsg_whenOnMsg_thenVerifyOutput_NoSendAttributesDeletedNotification_Not
...
+        config.setUseAttributesScopeTemplate(true);
+        config.setScope("$[scopeInData]");
+        nodeConfiguration = new TbNodeConfiguration(JacksonUtil.valueToTree(config));
+        node.init(ctx, nodeConfiguration);
+
+        final String data = "{\"TestAttribute_2\": \"humidity\", \"TestAttribute_3\": \"voltage\", \"scopeInData\": \" SeRvEr_ScOpE \"}";
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_ATTRIBUTES_REQUEST, deviceId, TbMsgMetaData.EMPTY, data);
+
+        node.onMsg(ctx, msg);
+
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgDeleteAttributesNodeTest.java:218]: missing verify in the same way as in the tests for save attributes node.

verify(telemetryService).deleteAndNotify(eq(ctx.getTenantId()), eq(msg.getOriginator()), eq(AttributeScope.SERVER_SCOPE),
                eq(keysToDelete), eq(false), eq(new TelemetryNodeCallback(ctx, msg)));

Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/telemetry/TbMsgDeleteAttributesNodeTest.java:218]: The main idea to test that deleteAndNotify will use correct scope.

**Review by ShvaykaD — APPROVED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNode.java [hunk 7]
```diff
@@ -154,11 +154,27 @@ private boolean checkNotifyDeviceMdValue(String notifyDeviceMdValue) {
         return StringUtils.isEmpty(notifyDeviceMdValue) || Boolean.parseBoolean(notifyDeviceMdValue);
     }
 
-    private AttributeScope getScope(String mdScopeValue) {
+    private AttributeScope getScope(TbMsg msg) {
+        if (config.isUseAttributesScopeTemplate()) {
+            try {
+                return AttributeScope.parseFrom(TbNodeUtils.processPattern(config.getScope(), msg));
+            } catch (Exception e) {
+                String mdScopeValue = msg.getMetaData().getValue(SCOPE);
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/telemetry/TbMsgAttributesNode.java:162]: Consider just adding a hint that scope can be set via metadata field. Customers than will do the needed transformations on their own.

**Comment by irynamatveieva** [comment]
Closed because it was decided to just add a hint that the scope can be set via the metadata field.

**Pull request closed without merging**